File: extras/rhubarb-for-spine/rhubarb-for-spine/AudioFileModel.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace rhubarb_for_spine {
	public class AudioFileModel : ModelBase {
		private readonly string filePath;
		private readonly ISet<MouthShape> extendedMouthShapes;
		private readonly SemaphoreSlim semaphore;
		private readonly Action<IReadOnlyCollection<MouthCue>> reportResult;
		private bool animatedPreviously;
		private CancellationTokenSource cancellationTokenSource;

		private AudioFileStatus _status;
		private string _actionLabel;
		private double? _animationProgress;

		public AudioFileModel(
			string name,
			string filePath,
			string displayFilePath,
			string dialog,
			ISet<MouthShape> extendedMouthShapes,
			bool animatedPreviously,
			SemaphoreSlim semaphore,
			Action<IReadOnlyCollection<MouthCue>> reportResult
		) {
			this.reportResult = reportResult;
			Name = name;
			this.filePath = filePath;
			this.extendedMouthShapes = extendedMouthShapes;
			this.animatedPreviously = animatedPreviously;
			this.semaphore = semaphore;
			DisplayFilePath = displayFilePath;
			Dialog = dialog;
			Status = animatedPreviously ? AudioFileStatus.Done : AudioFileStatus.NotAnimated;
		}

		public string Name { get; }
		public string DisplayFilePath { get; }
		public string Dialog { get; }

		public double? AnimationProgress {
			get { return _animationProgress; }
			private set {
				_animationProgress = value;
				OnPropertyChanged(nameof(AnimationProgress));

				// Hack, so that a binding to Status will refresh
				OnPropertyChanged(nameof(Status));
			}
		}

		public AudioFileStatus Status {
			get { return _status; }
			private set {
				_status = value;
				ActionLabel = _status == AudioFileStatus.Scheduled || _status == AudioFileStatus.Animating
					? "Cancel"
					: "Animate";
				OnPropertyChanged(nameof(Status));
			}
		}

		public string ActionLabel {
			get { return _actionLabel; }
			private set {
				_actionLabel = value;
				OnPropertyChanged(nameof(ActionLabel));
			}
		}

		public void PerformAction() {
			switch (Status) {
				case AudioFileStatus.NotAnimated:
				case AudioFileStatus.Done:
					StartAnimation();
					break;
				case AudioFileStatus.Scheduled:
				case AudioFileStatus.Animating:
					CancelAnimation();
					break;
				default:
					throw new ArgumentOutOfRangeException();
			}
		}

		private async void StartAnimation() {
			cancellationTokenSource = new CancellationTokenSource();
			Status = AudioFileStatus.Scheduled;
			try {
				await semaphore.WaitAsync(cancellationTokenSource.Token);
				Status = AudioFileStatus.Animating;
				try {
					var progress = new Progress(value => AnimationProgress = value);
					IReadOnlyCollection<MouthCue> mouthCues = await RhubarbCli.AnimateAsync(
						filePath, Dialog, extendedMouthShapes, progress, cancellationTokenSource.Token);
					animatedPreviously = true;
					Status = AudioFileStatus.Done;
					reportResult(mouthCues);
				} finally {
					AnimationProgress = null;
					semaphore.Release();
					cancellationTokenSource = null;
				}
			} catch (OperationCanceledException) {
				Status = animatedPreviously ? AudioFileStatus.Done : AudioFileStatus.NotAnimated;
			}
		}

		private void CancelAnimation() {
			cancellationTokenSource.Cancel();
		}

		private class Progress : IProgress<double> {
			private readonly Action<double> report;
			private double value;

			public Progress(Action<double> report) {
				this.report = report;
			}

			public void Report(double progress) {
				value = progress;
				report(progress);
			}
		}
	}

	public enum AudioFileStatus {
		NotAnimated,
		Scheduled,
		Animating,
		Done
	}
}

File: extras/rhubarb-for-spine/src/main/kotlin/com/rhubarb_lip_sync/rhubarb_for_spine/AudioFileModel.kt
package com.rhubarb_lip_sync.rhubarb_for_spine

import javafx.application.Platform
import javafx.beans.binding.BooleanBinding
import javafx.beans.binding.ObjectBinding
import javafx.beans.binding.StringBinding
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.scene.control.Alert
import javafx.scene.control.ButtonType
import tornadofx.getValue
import tornadofx.setValue
import java.util.concurrent.ExecutorService
import java.util.concurrent.Future

class AudioFileModel(
	audioEvent: SpineJson.AudioEvent,
	private val parentModel: AnimationFileModel,
	private val executor: ExecutorService,
	private val reportResult: (List<MouthCue>) -> Unit
) {
	val spineJson = parentModel.spineJson

	val audioFilePath = spineJson.audioDirectoryPath.resolve(audioEvent.relativeAudioFilePath)

	val eventNameProperty = SimpleStringProperty(audioEvent.name)
	val eventName by eventNameProperty

	val displayFilePathProperty = SimpleStringProperty(audioEvent.relativeAudioFilePath)
	val displayFilePath by displayFilePathProperty

	val animationNameProperty = SimpleStringProperty().apply {
		val mainModel = parentModel.parentModel
		bind(object : ObjectBinding<String>() {
			init {
				super.bind(
					mainModel.animationPrefixProperty,
					eventNameProperty,
					mainModel.animationSuffixProperty
				)
			}
			override fun computeValue(): String {
				return mainModel.animationPrefix + eventName + mainModel.animationSuffix
			}
		})
	}
	val animationName by animationNameProperty

	val dialogProperty = SimpleStringProperty(audioEvent.dialog)
	val dialog: String? by dialogProperty

	val animationProgressProperty = SimpleObjectProperty<Double?>(null)
	var animationProgress by animationProgressProperty
		private set

	private val animatedProperty = SimpleBooleanProperty().apply {
		bind(object : ObjectBinding<Boolean>() {
			init {
				super.bind(animationNameProperty, parentModel.spineJson.animationNames)
			}
			override fun computeValue(): Boolean {
				return parentModel.spineJson.animationNames.contains(animationName)
			}
		})
	}
	private var animated by animatedProperty

	private val futureProperty = SimpleObjectProperty<Future<*>?>()
	private var future by futureProperty

	private val audioFileStateProperty = SimpleObjectProperty<AudioFileState>().apply {
		bind(object : ObjectBinding<AudioFileState>() {
			init {
				super.bind(animatedProperty, futureProperty, animationProgressProperty)
			}
			override fun computeValue(): AudioFileState {
				return if (future != null) {
					if (animationProgress != null)
						if (future!!.isCancelled)
							AudioFileState(AudioFileStatus.Canceling)
						else
							AudioFileState(AudioFileStatus.Animating, animationProgress)
					else
						AudioFileState(AudioFileStatus.Pending)
				} else {
					if (animated)
						AudioFileState(AudioFileStatus.Done)
					else
						AudioFileState(AudioFileStatus.NotAnimated)
				}
			}
		})
	}
	private val audioFileState by audioFileStateProperty

	val busyProperty = SimpleBooleanProperty().apply {
		bind(object : BooleanBinding() {
			init {
				super.bind(futureProperty)
			}
			override fun computeValue(): Boolean {
				return future != null
			}

		})
	}
	val busy by busyProperty

	val statusLabelProperty = SimpleStringProperty().apply {
		bind(object : StringBinding() {
			init {
				super.bind(audioFileStateProperty)
			}
			override fun computeValue(): String {
				return when (audioFileState.status) {
					AudioFileStatus.NotAnimated -> "Not animated"
					AudioFileStatus.Pending -> "Waiting"
					AudioFileStatus.Animating -> "${((animationProgress ?: 0.0) * 100).toInt()}%"
					AudioFileStatus.Canceling -> "Canceling"
					AudioFileStatus.Done -> "Done"
				}
			}
		})
	}
	val statusLabel by statusLabelProperty

	val actionLabelProperty = SimpleStringProperty().apply {
		bind(object : StringBinding() {
			init {
				super.bind(futureProperty)
			}
			override fun computeValue(): String {
				return if (future != null)
					"Cancel"
				else
					"Animate"
			}
		})
	}
	val actionLabel by actionLabelProperty

	fun performAction() {
		if (future == null) {
			if (animated) {
				Alert(Alert.AlertType.CONFIRMATION).apply {
					headerText = "Animation '$animationName' already exists."
					contentText = "Do you want to replace the existing animation?"
					val result = showAndWait()
					if (result.get() != ButtonType.OK) {
						return
					}
				}
			}

			startAnimation()
		} else {
			cancelAnimation()
		}
	}

	private fun startAnimation() {
		val wrapperTask = Runnable {
			val extendedMouthShapes = parentModel.mouthShapes.filter { it.isExtended }.toSet()
			val reportProgress: (Double?) -> Unit = {
				progress -> runAndWait { this@AudioFileModel.animationProgress = progress }
			}
			val rhubarbTask = RhubarbTask(audioFilePath, dialog, extendedMouthShapes, reportProgress)
			try {
				try {
					val result = rhubarbTask.call()
					runAndWait {
						reportResult(result)
					}
				} finally {
					runAndWait {
						animationProgress = null
						future = null
					}
				}
			} catch (e: InterruptedException) {
			} catch (e: Exception) {
				Platform.runLater {
					Alert(Alert.AlertType.ERROR).apply {
						headerText = "Error performing lip sync for event '$eventName'."
						contentText = e.message
						show()
					}
				}
			}

		}
		future = executor.submit(wrapperTask)
	}

	private fun cancelAnimation() {
		future?.cancel(true)
	}
}

enum class AudioFileStatus {
	NotAnimated,
	Pending,
	Animating,
	Canceling,
	Done
}

data class AudioFileState(val status: AudioFileStatus, val progress: Double? = null)